### PR TITLE
[CORDA-2866] Prevent node startup if legal identity key is lost

### PR DESCRIPTION
**Pull request by JamesHR3**
If the node loses its legal identity key (from `nodekeystore.jks`), then the next time it starts up it will generate a new one and publish this to the network map. The network map currently cannot handle finding a new key for the same identity, so this results in a number of network issues. This PR prevents this by preventing the node from starting if there is no legal identity key in `nodekeystore.jks`, but there is a key associated with the node's X500 name in the identity store.

### CODE REVIEW

**Comment by JamesHR3** [comment]
Anthony suggested that we would be better off with an integration style test here, and I'm inclined to agree, so might be worth holding off on reviewing this until that is present.

File: node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt [hunk 1]
```diff
@@ -838,6 +838,16 @@ abstract class AbstractNode<S>(val configuration: NodeConfiguration,
...
         var signingCertificateStore = configuration.signingCertificateStore.get()
         if (!cryptoService.containsKey(legalIdentityPrivateKeyAlias) && !signingCertificateStore.contains(legalIdentityPrivateKeyAlias)) {
+            // Directly use the X500 name to public key map, as the identity service requires the node identity to start correctly.
+            database.transaction {
+                val x500Map = PersistentIdentityService.createX500Map(cacheFactory)
+                require(x500Map[configuration.myLegalName] == null) {
+                    // There is already a party in the identity store for this node, but the key has been lost. If this node starts up, it will
+                    // publish it's new key to the network map, which Corda cannot currently handle. To prevent this, stop the node from starting.
+                    "$legalIdentityPrivateKeyAlias not found in key store, but the corresponding X500 name ${configuration.myLegalName}" +
+                            " has a keypair. This suggests the identity for this node has been lost. Shutting down to prevent network issues."
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt]: network map issues

File: node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt [hunk 2]
```diff
@@ -838,6 +838,16 @@ abstract class AbstractNode<S>(val configuration: NodeConfiguration,
 
         var signingCertificateStore = configuration.signingCertificateStore.get()
         if (!cryptoService.containsKey(legalIdentityPrivateKeyAlias) && !signingCertificateStore.contains(legalIdentityPrivateKeyAlias)) {
+            // Directly use the X500 name to public key map, as the identity service requires the node identity to start correctly.
+            database.transaction {
+                val x500Map = PersistentIdentityService.createX500Map(cacheFactory)
+                require(x500Map[configuration.myLegalName] == null) {
+                    // There is already a party in the identity store for this node, but the key has been lost. If this node starts up, it will
+                    // publish it's new key to the network map, which Corda cannot currently handle. To prevent this, stop the node from starting.
+                    "$legalIdentityPrivateKeyAlias not found in key store, but the corresponding X500 name ${configuration.myLegalName}" +
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt]: For ENT the private key will not be in the key store if it's using an HSM. Probablysay "Private key for the node legal identity not found (alias $legalIdentityPrivateKeyAlias) but the public key for it exists in the database."

File: node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt [hunk 3]
```diff
@@ -838,6 +838,16 @@ abstract class AbstractNode<S>(val configuration: NodeConfiguration,
 
         var signingCertificateStore = configuration.signingCertificateStore.get()
         if (!cryptoService.containsKey(legalIdentityPrivateKeyAlias) && !signingCertificateStore.contains(legalIdentityPrivateKeyAlias)) {
+            // Directly use the X500 name to public key map, as the identity service requires the node identity to start correctly.
+            database.transaction {
+                val x500Map = PersistentIdentityService.createX500Map(cacheFactory)
+                require(x500Map[configuration.myLegalName] == null) {
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt]: `configuration.myLegalName !in x500Map`

File: node/src/integration-test/kotlin/net/corda/node/BootTests.kt [hunk 4]
```diff
@@ -64,6 +67,30 @@ class BootTests {
...
+        driver(DriverParameters(notarySpecs = emptyList(), inMemoryDB = false, startNodesInProcess = false)) {
+            val alice = startNode(providedName = ALICE_NAME).getOrThrow()
+            val aliceDir = alice.baseDirectory / "certificates"
+            (aliceDir / "nodekeystore.jks").toFile().delete()
+            val cert = CertificateStoreStubs.Signing.withCertificatesDirectory(aliceDir).get(true)
+            cert.registerDevSigningCertificates(ALICE_NAME)
+            (alice as OutOfProcess).process.destroyForcibly()
+            alice.stop()
+            val customOverrides = mapOf(
+                    "p2pAddress" to "localhost:${alice.rpcAddress.port}"
```
[the remark this answers is not in the document]
Comment (JamesHR3) [reply node/src/integration-test/kotlin/net/corda/node/BootTests.kt]: Hmm. I borrowed this line from an existing test that was trying to do the same thing - you're right that it is a bit odd! I'll update the original instance too

File: node/src/integration-test/kotlin/net/corda/node/BootTests.kt [hunk 5]
```diff
@@ -64,6 +68,26 @@ class BootTests {
...
     }
+
+    @Test
+    fun `node fails to start if identity key for x500 name is lost`() {
+        driver(DriverParameters(notarySpecs = emptyList(), inMemoryDB = false, startNodesInProcess = false)) {
+            val alice = startNode(providedName = ALICE_NAME).getOrThrow()
+            val aliceCertDir = alice.baseDirectory / "certificates"
+            (aliceCertDir / "nodekeystore.jks").delete()
+            val cert = CertificateStoreStubs.Signing.withCertificatesDirectory(aliceCertDir).get(true)
+            cert.registerDevSigningCertificates(ALICE_NAME)
```
Comment (shamsasari) [inline node/src/integration-test/kotlin/net/corda/node/BootTests.kt]: What is this line doing, it's not clear

Comment (JamesHR3) [reply node/src/integration-test/kotlin/net/corda/node/BootTests.kt]: Turns out creating the nodekeystore.jks file doesn't actually populate it with any keys or certificates, including the "cordaclientca" certificate. Without this, the node fails to start, but at an earlier point to the check I was trying to implement.

Comment (shamsasari) [reply node/src/integration-test/kotlin/net/corda/node/BootTests.kt]: I was going to suggest to add a comment explaining what's going on here, but I think it's better to simply rename this method as it's not clear. `installDevNodeCaCertPath`

File: node/src/integration-test/kotlin/net/corda/node/BootTests.kt [hunk 6]
```diff
@@ -64,6 +68,26 @@ class BootTests {
             assertEquals(1, numberOfNodesThatLogged)
         }
     }
+
+    @Test
+    fun `node fails to start if identity key for x500 name is lost`() {
```
Comment (shamsasari) [inline node/src/integration-test/kotlin/net/corda/node/BootTests.kt]: `node fails to start if its legal identity key is lost`

File: node/src/integration-test/kotlin/net/corda/node/BootTests.kt [hunk 7]
```diff
@@ -64,6 +68,26 @@ class BootTests {
...
+        driver(DriverParameters(notarySpecs = emptyList(), inMemoryDB = false, startNodesInProcess = false)) {
+            val alice = startNode(providedName = ALICE_NAME).getOrThrow()
+            val aliceCertDir = alice.baseDirectory / "certificates"
+            (aliceCertDir / "nodekeystore.jks").delete()
+            val cert = CertificateStoreStubs.Signing.withCertificatesDirectory(aliceCertDir).get(true)
+            cert.installDevNodeCaCertPath(ALICE_NAME)
```
Comment (shamsasari) [inline node/src/integration-test/kotlin/net/corda/node/BootTests.kt]: Can you add a comment why this is needed

**Pull request merged**